### ISSUE TITLE
Consider alternative, simpler and faster way to mock runtime in sdk

### ISSUE THREAD

**Issue by matklad** [near/near-sdk-rs#373]
TL;DR: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=833aa3518e2173f47b25508400b2934f

At the moment, we use a rather sophisticated blockchain interface to facilitate using both real wasm-based blockchain and a mock blockchaing for tests.

We store a dynamically dispatch blockchain interface in a thread local:

https://github.com/near/near-sdk-rs/blob/97fc632fcc58eb7ff7faad0be54ea8ec91dbf694/near-sdk/src/environment/env.rs#L17-L22

For the real blockchaing, if we are compiling for wasm, we include near's ABI here:

https://github.com/near/near-sdk-rs/blob/97fc632fcc58eb7ff7faad0be54ea8ec91dbf694/near-sdk-macros/res/sys.rs#L1-L9

And implement that dyn interface for it:

https://github.com/near/near-sdk-rs/blob/97fc632fcc58eb7ff7faad0be54ea8ec91dbf694/near-sdk-macros/res/near_blockchain.rs#L1-L16

Similarly, there's a mocked blockchain here:

https://github.com/near/near-sdk-rs/blob/97fc632fcc58eb7ff7faad0be54ea8ec91dbf694/near-sdk/src/environment/mocked_blockchain.rs#L1

The result is that even wasm contracts need to go through thread-local access and dynamic dispatch, which hurts runtime.

https://github.com/near/near-sdk-rs/blob/97fc632fcc58eb7ff7faad0be54ea8ec91dbf694/near-sdk/src/environment/env.rs#L693-L703

On the complexity front, the implementation of sdk needs to explicitly care about existence of the mock blockchain, it can't just directly use ABI exposed by near.

There's a simple way to do this! Instead of using `&dyn` to inject a mock blockchain, we can just *define* the `extern "C"` interface ourselves:

```rust
extern "C" {
    pub fn log_utf8(len: u64, ptr: u64);
}

#[cfg(not(target_arch = "wasm32"))]
mod mock_chain {
    #[no_mangle]
    extern "C" fn log_utf8(len: u64, ptr: u64) {
        unsafe {
            let s = std::slice::from_raw_parts(ptr as *const u8, len as usize);
            let s = std::str::from_utf8_unchecked(s);
            println!("{}", s);
        }
    }
}

pub fn log(s: &str) {
    unsafe {
        log_utf8(s.len() as u64, s.as_ptr() as *const u8 as u64)
    }
}

fn main() {
    log("Hello, world");
}
```

The `extern "C" {` block corresponds to sys.rs file. Note that we include it unconditionally, for all tragets. 

The `log` function corresponds to the SDK API. It is codded against near ABI and, when compiled to wasm, produces the optimal code without any extra logic to handle "is this a mock blockchan?"

The `mock_chain` module is included if we are not compiling for wasm. It *defines* the functions of the near ABI. When the code is compiled for host, the linker links these mock definitions with declarations in the `extern "C"` block. Note that mock_chain may even live in a separate crate. 

To clarify, the mock chain would probably still want to use therad-local storage for a blockthain struct, to allow the user to set VMContext and such. However, there's no dynamic dispatch anywhere, the SDK proper doesn't need to think about mocking, and the compiled wasm code is unaware of mocks as well. 

I've learned about this linker-based dependency injection technique from the rust standard library initially: https://github.com/rust-lang/rust/blob/88b99dec2a4d103b8153ca9300bf0fbebdf65bda/library/std/src/panicking.rs#L33-L52

**Comment by nearmax** [comment]
Mind blown.
